cpp/include/raft/core/device_resources_snmg.hpp

```diff
@@ -16,226 +16,99 @@
+  ~device_resources_snmg(){};
 
-    ncclGroupStart();
-    for (int rank = 0; rank < num_ranks; rank++) {
-      ncclComm_t& nccl_comm = raft::resource::get_nccl_comm(clique[rank]);
-      RAFT_NCCL_TRY_NO_THROW(ncclCommDestroy(nccl_comm));
-    }
-    ncclGroupEnd();
-  }
-
-  /**
-   * @brief Set root rank of NCCL clique
-   */
-  inline void set_nccl_root_rank(int rank)
-  {
-    raft::resource::set_nccl_clique_root_rank(*this, rank);
-  }
-
-  /**
-   * @brief Get root rank of NCCL clique
-   */
-  inline int get_nccl_root_rank() const { return raft::resource::get_nccl_clique_root_rank(*this); }
-
-  /**
-   * @brief Get number of ranks in NCCL clique
-   */
-  inline int get_nccl_num_ranks() const { return raft::resource::get_nccl_num_ranks(*this); }
-
-  /**
-   * @brief Get device ID of rank in NCCL clique
-   */
-  inline int get_device_id_of_rank(int rank) const
-  {
-    const raft::resources& dev_res = raft::resource::get_device_resources_for_rank(*this, rank);
-    return raft::resource::get_device_id(dev_res);
-  }
-
-  /**
-   * @brief Get NCCL comm object of rank in NCCL clique
-   */
-  inline ncclComm_t& get_nccl_comm_for_rank(int rank) const
-  {
-    return raft::resource::get_nccl_comm_for_rank(*this, rank);
-  }
-
-  /**
-   * @brief Get raft::resources object of rank in NCCL clique
-   */
-  inline const raft::resources& get_device_resources_for_rank(int rank) const
-  {
-    return raft::resource::get_device_resources_for_rank(*this, rank);
-  }
-
-  /**
-   * @brief Set current device ID to rank and return its raft::resources object
-   */
-  inline const raft::resources& set_current_device_to_rank(int rank) const
-  {
-    return raft::resource::set_current_device_to_rank(*this, rank);
-  }
-
-  /**
-   * @brief Set current device ID to root rank and return its raft::resources object
-   */
-  inline const raft::resources& set_current_device_to_root_rank() const
-  {
-    return raft::resource::set_current_device_to_root_rank(*this);
-  }
-
-  /**
-   * @brief Set a memory pool on all GPUs of the clique
-   */
-  void set_memory_pool(int percent_of_free_memory) const
-  {
-    for (int rank = 0; rank < get_nccl_num_ranks(); rank++) {
-      const raft::resources& dev_res = set_current_device_to_rank(rank);
-      // check limit for each device
-      size_t limit = rmm::percent_of_free_device_memory(percent_of_free_memory);
-      raft::resource::set_workspace_to_pool_resource(dev_res, limit);
-    }
-    cudaSetDevice(this->main_gpu_id_);
-  }
-
-  bool has_resource_factory(resource::resource_type resource_type) const override
-  {
-    if (resource_type != raft::resource::NCCL_CLIQUE &&
-        resource_type != raft::resource::NCCL_COMM &&
-        resource_type != raft::resource::CLIQUE_ROOT_RANK)
-      // for resources unrelated to SNMG switch current GPU to main GPU ID
-      cudaSetDevice(this->main_gpu_id_);
-    return raft::resources::has_resource_factory(resource_type);
-  }
+  int main_gpu_id_;
```

Comment:
Perhaps we should change 'root_rank' to this above. I like "main_gpu"id" better.

Reply:
Thinking of it, why do we need the `main_gpu_id_` within `device_resources_snmg` class? Can we just use the `get_root_rank` (or after changing it, the `get_main_gpu`) function to retrieve the main gpu id instead? @viclafargue

Reply:
Yes, ideally there's no internal state (e.g. member variables) being stored directly on the `device_resources` instances and all state is coming from the underlying `raft::resources` (and being accessed and mutated with `raft::resource::get_` and `raft::resources::set_`).

Reply:
`raft::resources` (and `raft::resource` are the underlying contracts between the user and the lower-level APIs in RAFT/cuVS/cuML/cuGraph/cuOpt), `raft::device_resources_...` is just a convenience API so users don't have to concern themselves with worrying about calling `raft::resource::set_xxxx` before they can call any of our functions.

Reply:
I've made it so that we call the `raft::resource::set_main_gpu_id` in the constructor of `device_resources_snmg`

Reply:
Perfect!

Reply:
The root rank (not device ID) is part of the clique (unlike the main GPU ID which may lie outside of it) and serves a specific role in the MG algorithm. For example, operations such as ANN merging can be performed on the root rank. In contrast, the main GPU is primarily used to enable the `device_resources_snmg` class to function as a single-GPU resource. It is tied to this utility class and is not directly involved in MG operations.